### PR TITLE
feat(medusa,pricing): Cart pricing context with customer group

### PR DESCRIPTION
**Pull request by riqwan**
what:

- cart workflows are updated to now accept cart context as pricing context
- pricing module is updated to handle nested rules
- `customer_group_id` rule is renamed `customer.groups.id` in line with cart shape

Thanks to initial contribution from @sergiocampama - https://github.com/medusajs/medusa/pull/10493

RESOLVES CMRC-745

### CODE REVIEW

**Comment by docloulou** [comment]
Well spotted ;)

**Comment by docloulou** [comment]
@riqwan same situation here [Order add line item WF](https://github.com/medusajs/medusa/blob/03803283d68b72c341c3fccf2827eaf04ebc063a/packages/core/core-flows/src/order/workflows/add-line-items.ts#L105C1-L118C6)

**Comment by riqwan** [comment]
@docloulou thanks, there are 10 different workflows in orders that needs to be updated, but that will come in a different PR. 🤞🏻

File: packages/modules/pricing/src/migrations/Migration20241212190401.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { Migration } from "@mikro-orm/migrations"
+
+export class Migration20241212190401 extends Migration {
+  async up(): Promise<void> {
+    this.addSql(
+      `UPDATE price_list_rule SET attribute = 'customer.groups.id' WHERE attribute = 'customer_group_id';`
```
Comment (olivermrbl) [inline packages/modules/pricing/src/migrations/Migration20241212190401.ts:6]: q: Why did we end up renaming the attribute? Feel like this is an awkward key to have

File: packages/core/core-flows/src/cart/workflows/update-line-item-in-cart.ts [hunk 2]
```diff
@@ -4,34 +4,43 @@ import {
...
 import { confirmVariantInventoryWorkflow } from "./confirm-variant-inventory"
 import { refreshCartItemsWorkflow } from "./refresh-cart-items"
 
+const cartFields = cartFieldsForPricingContext.concat(["items.*"])
+
 export const updateLineItemInCartWorkflowId = "update-line-item-in-cart"
 /**
  * This workflow updates a cart's line item.
  */
 export const updateLineItemInCartWorkflow = createWorkflow(
```
Comment (olivermrbl) [inline packages/core/core-flows/src/cart/workflows/update-line-item-in-cart.ts:25]: note to self: Input type change is a breaking change.

**Review by olivermrbl — COMMENTED** [review]
Overall LGTM, just a comment on the attribute change

File: packages/modules/pricing/src/migrations/Migration20241212190401.ts (hunk 1 above)
Comment (riqwan) [reply packages/modules/pricing/src/migrations/Migration20241212190401.ts:6]: The attributes on all our rules based engines are based on what the cart shape looks like - eg: `items.product.id`, `items.subtotal`, etc. This change brings it into alignment. Prevents us from having to mutate the pricing context wherever needed and relying on a singular shape. 

Important to note though that this won't be seen in the presentation layer, for the users they'd just see "Customer Groups"

**Review by olivermrbl — APPROVED** [review]
LGTM, the filter utilities are a bit daunting to review, so primarily looked at the tests, and they seem good :)

File: packages/core/utils/src/common/filter-object-by-keys.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import { isDefined } from "./is-defined"
+
+export function filterObjectByKeys(obj, paths) {
```
Comment (adrien2p) [inline packages/core/utils/src/common/filter-object-by-keys.ts:3]: suggestion: A description and example here would be nice to have

File: packages/core/utils/src/common/flatten-object-to-key-value-pairs.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
+type NestedObject = {
+  [key: string]: any
+}
+
+export function flattenObjectToKeyValuePairs(obj: NestedObject): NestedObject {
```
Comment (adrien2p) [inline packages/core/utils/src/common/flatten-object-to-key-value-pairs.ts:5]: same suggestion as the previous one

File: packages/core/utils/src/common/flatten-object-to-key-value-pairs.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        if (item && typeof item === "object") {
+          Object.keys(item as object).forEach((k) => keys.add(k))
+        }
+      })
+
+      // Process each key
+      keys.forEach((key) => {
+        const values = arrayObjects
+          .map((item) => {
+            if (item && typeof item === "object") {
```
Comment (adrien2p) [inline packages/core/utils/src/common/flatten-object-to-key-value-pairs.ts:90]: suggestion: for those type of check would the util `isObject` be more convenient?

File: packages/core/utils/src/common/flatten-object-to-key-value-pairs.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+              return (item as Record<string, unknown>)[key]
+            }
+            return undefined
+          })
+          .filter((v) => v !== undefined)
+
+        if (values.length > 0) {
+          const newPath = `${pathStr}.${key}`
+          if (values.every((v) => typeof v === "object" && !Array.isArray(v))) {
```
Comment (adrien2p) [inline packages/core/utils/src/common/flatten-object-to-key-value-pairs.ts:99]: suggestion: should we use isObject?

**Review by adrien2p — APPROVED** [review]

File: packages/core/utils/src/common/flatten-object-to-key-value-pairs.ts (hunk 6 above)
Comment (riqwan) [reply packages/core/utils/src/common/flatten-object-to-key-value-pairs.ts:99]: @adrien2p sorry, just saw this. Will open a new PR with these changes to get this PR into the release.

**Comment by sergiocampama** [comment]
thanks @riqwan! do you know when a new version might get released with this fix?

**Comment by riqwan** [comment]
@sergiocampama today!

**Pull request merged**